File: Kalorie/Kalorie/Core/Models/FoodItemValidation.swift
//
//  FoodItemValidation.swift
//  Kalorie
//
//  Created by Josef Antoni on 10.09.2026.
//

import Foundation

enum FoodItemValidationError: Error {
    case invalidCode
    case invalidName
    case invalidCalories
    case invalidWeight
    case invalidPortion(FoodPortionError)
}

extension FoodItemValidationError {

    // MARK: - Functions

    func mapped<T>(
        invalidCode: T,
        invalidName: T,
        invalidCalories: T,
        invalidWeight: T,
        invalidPortion: (FoodPortionError) -> T
    ) -> T {
        switch self {
        case .invalidCode: return invalidCode
        case .invalidName: return invalidName
        case .invalidCalories: return invalidCalories
        case .invalidWeight: return invalidWeight
        case .invalidPortion(let portionError): return invalidPortion(portionError)
        }
    }
}

enum FoodItemValidation {

    // MARK: - Functions

    static func isValidBarcode(_ id: String) -> Bool {
        id.allSatisfy { $0.isASCII && $0.isNumber } && [8, 12, 13].contains(id.count)
    }

    static func isValidSubmissionUUID(_ id: String) -> Bool {
        UUID(uuidString: id) != nil && id == id.uppercased()
    }

    static func validate(_ item: FoodItemDomain) -> FoodItemValidationError? {
        guard isValidBarcode(item.id) || isValidSubmissionUUID(item.id) else { return .invalidCode }
        guard !item.czName.isEmpty else { return .invalidName }
        guard item.caloriesPerHundredGrams > 0 else { return .invalidCalories }
        guard item.weight > 0 else { return .invalidWeight }
        for portion in item.portions {
            if let error = FoodPortionValidation.validate(name: portion.name, grams: portion.grams) {
                return .invalidPortion(error)
            }
        }
        if let error = FoodPortionValidation.validate(portions: item.portions) {
            return .invalidPortion(error)
        }
        return nil
    }
}

File: Android/app/src/main/java/antoni/kalorie/core/models/FoodItemValidation.kt
package antoni.kalorie.core.models

sealed class FoodItemValidationError : Exception() {
    data object InvalidCode : FoodItemValidationError()
    data object InvalidName : FoodItemValidationError()
    data object InvalidCalories : FoodItemValidationError()
    data object InvalidWeight : FoodItemValidationError()
    data class InvalidPortion(val error: FoodPortionError) : FoodItemValidationError()

    // MARK: - Functions

    fun <T> mapped(
        invalidCode: T,
        invalidName: T,
        invalidCalories: T,
        invalidWeight: T,
        invalidPortion: (FoodPortionError) -> T,
    ): T = when (this) {
        InvalidCode -> invalidCode
        InvalidName -> invalidName
        InvalidCalories -> invalidCalories
        InvalidWeight -> invalidWeight
        is InvalidPortion -> invalidPortion(error)
    }
}

object FoodItemValidation {

    // MARK: - Properties

    private val barcodeLengths = setOf(8, 12, 13)
    private val submissionUUIDPattern = Regex("[0-9A-F]{8}-[0-9A-F]{4}-[0-9A-F]{4}-[0-9A-F]{4}-[0-9A-F]{12}")

    // MARK: - Functions

    fun isValidBarcode(id: String): Boolean = id.all { it in '0'..'9' } && id.length in barcodeLengths

    fun isValidSubmissionUUID(id: String): Boolean = submissionUUIDPattern.matches(id)

    fun validate(item: FoodItemDomain): FoodItemValidationError? {
        if (!isValidBarcode(item.id) && !isValidSubmissionUUID(item.id)) return FoodItemValidationError.InvalidCode
        if (item.czName.isEmpty()) return FoodItemValidationError.InvalidName
        if (item.caloriesPerHundredGrams <= 0) return FoodItemValidationError.InvalidCalories
        if (item.weight <= 0) return FoodItemValidationError.InvalidWeight
        for (portion in item.portions) {
            FoodPortionValidation.validate(portion.name, portion.grams)?.let {
                return FoodItemValidationError.InvalidPortion(it)
            }
        }
        FoodPortionValidation.validate(item.portions)?.let { return FoodItemValidationError.InvalidPortion(it) }
        return null
    }
}
